### DISCUSSION TITLE
nidigital // session.create_source_waveform // Error: There are no pins specified in the pin list

### DISCUSSION THREAD

**Discussion by ChrisOKay** [ni/nimi-python#1665] [Q&A] [answered]
I am successfully using nidigital for quite a while now with simple patterns hardcoded in 1's and 0's.
Now I wanted to try to source external data to my dut.
After having read the docs, I understood that I had to:

1. start a session
2. load pin_map
3. create a waveform via `session.create_source_waveform_serial(wave_name, BROADCAST, 1, LSB)`
4. write waveform data via session.write_source_waveform_broadcast(wave_name, wave_data)
5. load timing which implements op_codes _source_start(wave_name)_ and _source_ to feed the data to the dut

However, when trying to execute step 3, I get the following error:
`nidigital.errors.DriverError: -1074118655: There are no pins specified in the pin list.`

I do not understand why there should be no pins specified? I already loaded the pin_map successfully, and I applied the method deliberatly on all pins. Trying to reload the pin_map before this method only results in "The pin map has already been configured.".

=== Update ===
After having played around with this a bit more, trying to apply this method to an "all-pins"  pins['AllPins'] group, I got the error
`The number of channels specified for serial source or capture memory cannot exceed one per site.`
So it seems like my initial understanding to source the waveform serially to multiple pins does not work the way I thought it would.

Applying the method to a single pin (pins['NameOfPin']) does work as expected.
So now it's more a question of:
Is "There are no pins [...]" a reasonable error message for wrongfully (?) calling a create_source_serial method on the session instead of correctly (?) a single pin?

**Comment by sbethur** [comment, marked as answer]
@ChrisOKay, I agree that we could improve the error message to better inform the users on what is wrong and what steps should be taken.

Can you please provide the code snippet with which you got the error so I can file an internal bug against NI-Digital driver?

**Reply by ChrisOKay** [reply]
That should be a minimal example illustrating both the informative and misleading error messages.
The files can be found in the attached zip as well (link at bottom).

create_source_error_minimal_example.py:
```
    import nidigital
    with nidigital.Session(resource_name="PXIe_6570") as session:
        session.load_pin_map('my.pinmap')
        wave_cfg = {
            'waveform_name': 'my_wave',
            'data_mapping': nidigital.SourceDataMapping.BROADCAST,
            'sample_width': 1,
            'bit_order': nidigital.BitOrder.LSB,
        }
        print('Calling create_source on specific pin works.')
        session.pins['SYNC1'].create_source_waveform_serial(**wave_cfg)
        print('###')
        print('calling create_source on more than one pin raises expected error message:')
        try:
            session.pins['SYNC1', 'SYNC2'].create_source_waveform_serial(**wave_cfg)
        except nidigital.errors.DriverError as error:
            print(error)
        print('###')
        print('calling create_source on session raises misleading error message:')
        session.create_source_waveform_serial(**wave_cfg)
```

my.pinmap:
```
<?xml version="1.0" encoding="utf-8"?>
<PinMap schemaVersion="1.2" xmlns="http://www.ni.com/TestStand/SemiconductorModule/PinMap.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<Instruments>
		<NIDigitalPatternInstrument group="Digital" name="PXIe_6570" numberOfChannels="32"/>
	</Instruments>
	<Pins>
		<DUTPin name="SYNC1"/>
		<DUTPin name="SYNC2"/>
	</Pins>
	<Relays/>
	<RelayGroups/>
	<RelayConfigurations/>
	<Sites>
		<Site siteNumber="0"/>
	</Sites>
	<Connections>
		<Connection channel="1" instrument="PXIe_6570" pin="SYNC1" siteNumber="0"/>
		<Connection channel="26" instrument="PXIe_6570" pin="SYNC2" siteNumber="0"/>
	</Connections>
</PinMap>
```
[create_source_error_minimal_example.zip](https://github.com/ni/nimi-python/files/7565931/create_source_error_minimal_example.zip)
